config_db.go

```diff
@@ -136,95 +136,169 @@ func byIDs(ids []string) bson.M {
+// overrides config document which is fetched from the local database and, if [includeOverrides],
+// the overrides in the overrides config document are applied to the rest of the configuration.
 func getSectionsBSON(ctx context.Context, ids []string, includeOverrides bool) ([]bson.Raw, error) {
-	missingIDs := ids
+	db := GetEnvironment().DB()
+	if sharedDB := GetEnvironment().SharedDB(); sharedDB != nil {
+		db = sharedDB
+	}
+
+	cur, err := db.Collection(ConfigCollection).Find(ctx, byIDs(ids))
+	if err != nil {
+		return nil, errors.Wrap(err, "finding local configuration sections")
+	}
 	docs := make([]bson.Raw, 0, len(ids))
-	if includeOverrides {
-		cur, err := GetEnvironment().DB().Collection(ConfigCollection).Find(ctx, byIDs(ids))
+	if err := cur.All(ctx, &docs); err != nil {
+		return nil, errors.Wrap(err, "iterating cursor for local configuration sections")
+	}
+
+	if GetEnvironment().SharedDB() != nil {
+		docs, err = overrideConfig(ctx, docs, includeOverrides)
 		if err != nil {
-			return nil, errors.Wrap(err, "finding local configuration sections")
+			return nil, errors.Wrap(err, "overriding config")
 		}
+	}
+
+	return docs, nil
+}
 
-		if err := cur.All(ctx, &docs); err != nil {
-			return nil, errors.Wrap(err, "iterating cursor for local configuration sections")
+// overrideConfig overrides the content of the overrides document with the contents of the overrides
+// document from the local database. If [applyOverrides] is true then overrides are applied to
+// all the supplied docs.
+func overrideConfig(ctx context.Context, docs []bson.Raw, applyOverrides bool) ([]bson.Raw, error) {
+	res := GetEnvironment().DB().Collection(ConfigCollection).FindOne(ctx, byId(overridesSectionID))
+	if err := res.Err(); err != nil {
+		if errors.Is(err, mongo.ErrNoDocuments) {
+			return docs, nil
 		}
+		return nil, errors.Wrap(err, "getting overrides document")
+	}
+	rawOverrides, err := res.Raw()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting raw overrides config")
+	}
+	var overrides OverridesConfig
+	if err := bson.Unmarshal(rawOverrides, &overrides); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling overrides config")
+	}
 
-		var docIDs []string
-		for _, doc := range docs {
-			id, err := doc.LookupErr("_id")
-			if err != nil {
-				continue
-			}
-			idString, ok := id.StringValueOK()
-			if !ok {
-				continue
-			}
-			docIDs = append(docIDs, idString)
+	newDocs := make([]bson.Raw, 0, len(docs))
+	for _, doc := range docs {
+		newDoc, err := overrideDoc(doc, rawOverrides, overrides, applyOverrides)
+		if err != nil {
+			return nil, errors.Wrap(err, "overriding config section")
 		}
+		newDocs = append(newDocs, newDoc)
+	}
+
+	return newDocs, nil
+}
 
-		missingIDs, _ = utility.StringSliceSymmetricDifference(ids, docIDs)
+func overrideDoc(doc, rawOverrides bson.Raw, overrides OverridesConfig, applyOverrides bool) (bson.Raw, error) {
```

Comment:
Also not sure what the difference of doc and rawOverrides is. Reading it, if the doc is the same id as the section it returns the rawOverrides but if not it returns the doc (modified if applyOverrides is true)? Could you add comments to this one and change the names a bit

Reply:
The function has been refactored for clarity.